### PR TITLE
TTY: Implement line editing in canonical mode

### PR DESCRIPTION
**Pull request by j-piecuch**

### CODE REVIEW

File: sys/kern/tty.c [hunk 1]
```diff
@@ -173,6 +216,123 @@ static void tty_outq_putc(tty_t *tty, uint8_t c) {
...
+ */
+static void tty_echo(tty_t *tty, uint8_t c) {
+  assert(mtx_owned(&tty->t_lock));
+
+  int lflag = tty->t_lflag;
+  /* If ECHO is not set, only echo newlines provided ECHONL is set. */
+  if (!(lflag & ECHO) && (!(lflag & ECHONL) || c != '\n'))
+    return;
+  /* If ECHOCTL is set, echo control characters as ^A, ^B etc. */
+  if ((lflag & ECHO) && ((c <= 0x1f && c != '\t' && c != '\n') || c == 0x7f)) {
```
Comment (cahirwpz) [inline sys/kern/tty.c:231]: Two magic constants... is there any macro that defines them?

File: sys/kern/tty.c [hunk 2]
```diff
@@ -173,6 +216,123 @@ static void tty_outq_putc(tty_t *tty, uint8_t c) {
...
+    tty_output(tty, '^');
+    if (c == 0x7f)
```
Comment (cahirwpz) [inline sys/kern/tty.c:233]: ditto

File: sys/kern/tty.c [hunk 3]
```diff
@@ -191,16 +353,74 @@ void tty_input(tty_t *tty, uint8_t c) {
...
-      tty_output(tty, CTRL('g'));
-      tty_notify_out(tty);
+  if (lflag & ICANON) {
+    /* Canonical mode character processing takes place here. */
+    if (CCEQ(cc[VERASE], c)) {
+      /* Erase/backspace */
+      uint8_t erased;
+      if (tty_line_unputc(tty, &erased))
+        tty_erase(tty, erased);
+      goto notify;
```
Comment (cahirwpz) [inline sys/kern/tty.c:363]: Try to refrain yourself from using `goto`'s in cases where control flow is already complex.

**Review by cahirwpz — COMMENTED** [review]
Introduced control flow is very complex and difficult to read. Please do everything you can to keep it simple to read and understand.

File: sys/kern/tty.c (hunk 2 above)
Comment (j-piecuch) [reply sys/kern/tty.c:233]: There's no meaning to this constant apart from "the character 0x7f". For this reason, I don't think using a macro here makes sense. If I were to use a macro here, what would it be called, `CHAR_7F`?

File: sys/kern/tty.c (hunk 1 above)
Comment (j-piecuch) [reply sys/kern/tty.c]: I have added one.

File: sys/kern/tty.c (hunk 3 above)
Comment (j-piecuch) [reply sys/kern/tty.c]: Done.

File: sys/kern/tty.c (hunk 2 above)
Comment (cahirwpz) [reply sys/kern/tty.c:236]: Isn't that a code for delete character?

Comment (j-piecuch) [reply sys/kern/tty.c]: It is. I have added the macro.

**Review by cahirwpz — APPROVED** [review]
LGTM

**Pull request merged**